src/main/flight/failsafe.c

```diff
@@ -159,50 +171,103 @@ void failsafeUpdateState(void)
+                            failsafeState.throttleLowCounter = 5 * failsafeConfig->failsafe_throttle_low_delay;
+                        }
+                    }
+                    if (!receivingRxData) {
+                        if (failsafeSwitchIsOn && failsafeConfig->failsafe_kill_switch) {
+                            // KillswitchEvent: failsafe switch is configured as KILL switch and is switched ON
+                            failsafeActivate();
+                            failsafeState.phase = FAILSAFE_LANDED;      // skip auto-landing procedure
+                            failsafeState.receivingRxDataCounts = 0;    // require 0 seconds of valid rxData at 50Hz
+                        } else if (!feature(FEATURE_3D) && !failsafeState.throttleLowCounter) {
+                            // JustDisarm: not in 3D mode and throttle command was below 'min_throttle' for at least 'failsafe_throttle_low_delay' seconds
+                            failsafeActivate();
+                            failsafeState.phase = FAILSAFE_LANDED;      // skip auto-landing procedure
+                            failsafeState.receivingRxDataCounts = 150;  // require 3 seconds of valid rxData at 50Hz
+                        } else {
+                            failsafeState.phase = FAILSAFE_RX_LOSS_DETECTED;
+                        }
+                        reprocessState = true;
+                    }
+                } else {
+                    // When NOT armed, show state of failsafe switch in GUI (failsafe mode)
+                    if (failsafeSwitchIsOn) {
+                        ENABLE_FLIGHT_MODE(FAILSAFE_MODE);
+                    } else {
+                        DISABLE_FLIGHT_MODE(FAILSAFE_MODE);
+                    }
+                    // Keep counter cleared
+                    failsafeState.throttleLowCounter = 0;
                 }
                 break;
 
             case FAILSAFE_RX_LOSS_DETECTED:
-
+                if (receivingRxData) {
+                    failsafeState.phase = FAILSAFE_RX_LOSS_RECOVERED;
+                    reprocessState = true;
+                }
                 if (failsafeShouldForceLanding(armed)) {
                     // Stabilize, and set Throttle to specified level
                     failsafeActivate();
-
                     reprocessState = true;
                 }
                 break;
 
             case FAILSAFE_LANDING:
+                if (receivingRxData) {
+                    failsafeState.phase = FAILSAFE_RX_LOSS_RECOVERED;
+                    reprocessState = true;
+                }
                 if (armed) {
                     failsafeApplyControlInput();
                     beeperMode = BEEPER_RX_LOST_LANDING;
                 }
-
                 if (failsafeShouldHaveCausedLandingByNow() || !armed) {
-
+                    failsafeState.receivingRxDataCounts = 1500; // require 30 seconds of valid rxData at 50Hz
```

Comment:
magic number alert!  use a constant and if possible build it using other constants.

`#define MY_CONSTANT (30 * RX_FREQUENCY)`

using a counter is introducing another bug that needs to be fixed because of the update rate of rx is not always 50hz.  See other FIXMEs.  Use a fixed time `shouldDoSomethingAt = ...` or record a start time and use an interval that can be extended.  `somethingHappenedAt = ...;   interval = 30_SECONDS`  ...  `if (something) interval += 5_SECONDS`.

or something.